### PR TITLE
rfc: clean it up and optimize

### PR DESCRIPTION
**Pull request by pulsejet**
My intention is to rip off everything that isn't or is hard to maintain, and hopefully make this more usable.

1. Remove ethernet face 
2. Remove face events
3. Remove the UI
4. Remove NDNLP reliability and retransmission
5. Completely rip off the ndn library here, switch to go-ndn instead

Various optimizations
1. Switch to `sync.Map`
2. GC optimizations
3. Remove pointless goroutines

Benchmark
1. `GOMAXPROCS=4 GOGC=400 ./yanfd` with one instance of putchunks and 2 instances of catchunks.
1. Goodput goes from ~3gbps to ~4.5gbps

### CODE REVIEW

**Comment by zjkmxy** [comment]
Thank you for your help. This relieves the maintenance burden. But I want to make sure that:
- What is the goal of YaNFD?
- Do we still want to support the Windows App?

File: face/unix-stream-transport.go [hunk 1]
```diff
@@ -8,36 +8,41 @@
...
 		return nil, core.ErrNotCanonical
 	}
 
 	t := new(UnixStreamTransport)
-	t.makeTransportBase(remoteURI, localURI, PersistencyPersistent, ndn.Local, ndn.PointToPoint, tlv.MaxNDNPacketSize)
+	t.makeTransportBase(remoteURI, localURI, PersistencyPersistent, ndn_defn.Local, ndn_defn.PointToPoint, ndn_defn.MaxNDNPacketSize)
 
 	// Set connection
 	t.conn = conn.(*net.UnixConn)
+	t.reader = bufio.NewReaderSize(t.conn, 32*ndn_defn.MaxNDNPacketSize)
```
Comment (zjkmxy) [inline face/unix-stream-transport.go:43]: How you decided `32*ndn_defn.MaxNDNPacketSize`? Just want something large enough?

File: fw/thread.go [hunk 2]
```diff
@@ -28,21 +28,21 @@ var Threads map[int]*Thread
 var LOCALHOST = []byte{0x6c, 0x6f, 0x63, 0x61, 0x6c, 0x68, 0x6f, 0x73, 0x74}
 
 // HashNameToFwThread hashes an NDN name to a forwarding thread.
-func HashNameToFwThread(name enc.Name) int {
+func HashNameToFwThread(name *enc.Name) int {
```
Comment (zjkmxy) [inline fw/thread.go:31]: `enc.Name` is already a slice:

https://github.com/zjkmxy/go-ndn/blob/ec49b726e8ddecf63698832b53f01f2ec81aa72b/pkg/encoding/name_pattern.go#L12

File: table/fib-strategy-tree.go [hunk 3]
```diff
@@ -80,8 +80,9 @@ func (f *fibStrategyTreeEntry) findLongestPrefixEntryEnc(name enc.Name) *fibStra
 func (f *FibStrategyTree) fillTreeToPrefixEnc(name enc.Name) *fibStrategyTreeEntry {
 	curNode := f.root.findLongestPrefixEntryEnc(name)
 	for depth := curNode.depth + 1; depth <= len(name); depth++ {
+		component, _ := enc.ComponentFromBytes(At(name, depth-1).Bytes())
```
Comment (zjkmxy) [inline table/fib-strategy-tree.go:83]: What is the purpose of this line?

**Review by yoursunny — CHANGES_REQUESTED** [review]
Ethernet face was last maintained in 2024-Aug and is not broken so it should be kept.
No objection in deleting other features.

File: face/unix-stream-transport.go (hunk 1 above)
Comment (pulsejet) [reply face/unix-stream-transport.go:43]: Yeah it's arbitrary. It's a *minimum* anyway, the buffer could be larger

File: table/fib-strategy-tree.go (hunk 3 above)
Comment (pulsejet) [reply table/fib-strategy-tree.go:83]: It's  a deep copy. I'm not really sure what is going on here: the underlying array gets deallocated / overwritten somehow if this isn't copied and it becomes garbage. If I remove the copies in this file:

```
root@0037b98ec2ac:/work/YaNFD# nfdc strategy set /svs/abcdefghijklmnopqrstuvwxyz /localhost/nfd/strategy/multicast
strategy-set prefix=/svs/abcdefghijklmnopqrstuvwxyz strategy=/localhost/nfd/strategy/multicast/v=1
root@0037b98ec2ac:/work/YaNFD# nfdc strategy set /svs/abcdef /localhost/nfd/strategy/multicast
strategy-set prefix=/svs/abcdef strategy=/localhost/nfd/strategy/multicast/v=1
root@0037b98ec2ac:/work/YaNFD# nfdc strategy list
prefix=/ strategy=/localhost/nfd/strategy/best-route/v=1
prefix=/%85h%0E/abcdef strategy=/localhost/nfd/strategy/multicast/v=1
prefix=/%85h%0E/abcdefk%27%07%25%08%09localhost%08%03nfd strategy=/tegy%08%09mul/cas/%20J%07%9F%D4t%E9%03/I%9C%F4Yr%99%A2oT/v=1
root@0037b98ec2ac:/work/YaNFD# 
```

**Comment by pulsejet** [comment]
Thanks for reviewing, Xinyu.

> * What is the goal of YaNFD?

I've more stuff planned for the future (which might not materialize). Essentially I want YaNFD to be the out-of-the box NDN daemon that's focuses on ease of use and is maintanable. I'll elaborate in a Saturday call if you're available.

> * Do we still want to support the Windows App?

For now, I don't see the point. Some of the stuff that is removed here can be pulled back in later. My priority for this PR is to nuke the `YaNFD/ndn` package.

**Comment by pulsejet** [comment]
> Ethernet face was last maintained in 2024-Aug and is not broken so it should be kept.

If it works then it's probably fine; I'll try to pull this back in.

**Comment by Pesa** [comment]
> If it works then it's probably fine; I'll try to pull this back in.

https://www.youtube.com/watch?v=UPw-3e_pzqU

File: table/fib-strategy-tree.go (hunk 3 above)
Comment (zjkmxy) [reply table/fib-strategy-tree.go:83]: Thank you for clarifying. This seems to be a big issue. Maybe we should add a comment explaining this.

**Comment by zjkmxy** [comment]
> > * Do we still want to support the Windows App?
> 
> For now, I don't see the point. Some of the stuff that is removed here can be pulled back in later. My priority for this PR is to nuke the `YaNFD/ndn` package.

Then let it die 🙈
~~TBH I'm kind of tired of satisfying Windows store's policy, which seems to be more strict than COVID days~~

> If it works then it's probably fine; I'll try to pull this back in.

Let's do it in a separate PR. This is already too large and it seems to be working.

**Pull request merged**